syncapi/storage/syncserver.go

```diff
@@ -253,9 +259,14 @@ func (d *SyncServerDatasource) addPDUDeltaToResponse(
 		return nil, err
 	}
 
+	// len(deltas) should be a good enough initial capacity
 	joinedRoomIDs := make([]string, 0, len(deltas))
+
 	for _, delta := range deltas {
-		joinedRoomIDs = append(joinedRoomIDs, delta.roomID)
+		if delta.membership == membershipJoin {
```

Comment:
getStateDeltas states that it may produce duplicate join events. Is it fine to pass multiple stateDeltas for the same room down `/sync`?

https://github.com/matrix-org/dendrite/blob/29841bed6b1a88787211368e6052a87a658c5714/syncapi/storage/syncserver.go#L723-L731

Seems like `getStateDeltas` is currently incomplete and some of this work could be moved into there instead?

Reply:
> getStateDeltas states that it may produce duplicate join events. Is it fine to pass multiple stateDeltas for the same room down /sync?

It's not fine, indeed. There will be duplicate EDUs added to the /sync response.

> Seems like getStateDeltas is currently incomplete and some of this work could be moved into there instead?

It seems it'd be more efficient if we just return `joinedRoomIDs` from `getStateDeltas`, so I think that is worth a change. This will also fix the duplication above. But it isn't really related to the completeness of the function I guess?

Reply:
No, but we can fix that function in another PR I suppose.

Reply:
In fact the duplicate join events will not only affect EDUs, but also affect PDUs I believe; there'll be no visual effects though. Duplicate join events will cause the joined room response for a room to be calculated multiple times in [`addRoomDeltaToResponse`](https://github.com/matrix-org/dendrite/blob/0f0941ba25e64c4a3784f9b746b4f2692c242127/syncapi/storage/syncserver.go#L575).

Reply:
I think Riot (and possibly other clients) will display that a user has joined a room multiple times, but it's not the end of the world.